text/0000-inline-buildpack.md

```diff
@@ -0,0 +1,126 @@
+- a `buildpack.toml` using the `id` and `version` from the buildpack entry in `project.toml`
+
+When the ephemeral buildpack is executed, its working directory will contain the artifacts in the app repo. In this way, an `inline` script MAY reference other scripts in the repo. For example, a `lib/utils.sh` may be sourced by an inline Bash script.
+
+## Example
+
+The following `project.toml` creates a buildpack that will run two Rake tasks after running JVM and Ruby buildpacks.
+
+```toml
+[project]
+id = "my-app"
+
+[[build.buildpacks]]
+id = "example/jvm"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "example/ruby"
+version = "1.0"
+
+[[build.buildpacks]]
+api = "0.3"
+id = "me/rake-tasks"
+inline = """
+  rake war
+  rake db:migrate
+"""
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It discourages buildpack reuse (i.e. if it's really easy to create a one-off buildpack, people are less like to make reuseable buildpacks and share them)
+- It encourages the copy-paste problems created by `Dockerfile`
+- There is no built-in mechanism to provide dependencies in the build plan. However, this can be worked-around with a generic build-plan-buildpack that either statically or dynamically creates entries in the build plan.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- https://github.com/buildpacks/rfcs/pull/17
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Prior Art
+[prior-art]: #prior-art
+
+- `Dockerfile`
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Can an inline buildpack be published to a registry?
+- Can inline buildpacks be root buildpacks?
+- Can a `project.toml` contain more than one inline buildpack?
+- How will platforms implement this behavior?
+    - A future `/lifecycle/prepare` could generate the temporary buildpack from the inline script.
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+The [Project Descriptor extension](https://github.com/buildpacks/spec/blob/master/extensions/project-descriptor.md) will be ammended to include:
+
+## [[build.buildpacks]]
+
+The build table MAY contain an array of buildpacks. The schema for this table is:
+
+```toml
+[[build.buildpacks]]
+api = "<buildpack api>"
+shell = "<string (optional default=/bin/sh)>"
+inline = "<script contents (optional)>"
+```
+
+This defines the buildpacks that a platform should use on the repo.
+
+Either an `id` or a `uri` MUST be included, but MUST NOT include both. If `uri` is provided the `version`, `inline`, and `shell` MUST NOT be allowed.
+
+The `api` and `inline` key MUST be used together (i.e. one is not valid without the other). When any of the keys `api`, `inline`, or `shell` are provide the `version` and `uri` MUST NOT be allowed.
+
+When an `inline` script is provided, the script will be executed using the app directory as the buildpack root directory.
```

Comment:
As is `CNB_BUILDPACK_DIR` will be set to `<app-dir>` rather than to `<buildpacks-dir>/<id>/<version>`? I think I agree with this but just want to clarify.

Reply:
I think this may require an additional key in `buildpack.toml` so that the detect and build phases know to do this.

Reply:
@ekcasey i agree that there needs to be a way for the buildpack to tell the lifecycle where the dir is (or that it's an inline buildpack), but i worry about a key like that `buildpack.toml` because i can't think of reason it would configurable by a buildpack author. 

Is there some other metadata file we could use to indicate this?

Reply:
I'm wondering if the platform can define the dir in `order.toml` (`--order`) for detect, and the lifecycle would pass it on to `group.toml` for build.

Reply:
I originally thought supporting `inline` in `buildpack.toml` was a good idea **but** it should be a separate RFC. However, if we include that in this RFC we could solve this problem and a separate problem preserving the value of `shell`.

For example: if, instead of writing a `bind/build` and `bin/detect` script for each inline buildpack, we wrote a `buildpack.toml` with an `inline` and a `shell` field, the lifecycle could run those buildpacks appropriately.

Adding a change to the `order.toml` schema to work around this use case when we may want to remove it in favor of `buildpack.toml` changes shortly thereafter doesn't seem ideal.

Reply:
Honestly, after thinking about this more, how important is it that `CNB_BUILDPACK_DIR` is the app dir? It does make sense on some level... but given that the working dir is the app dir and these are intended to be simple scripts, would a user realistically be using `CNB_BUILDPACK_DIR` instead of `PWD`?

Reply:
@ekcasey yea, i think you're right. The most important thing is that a line like `source lib/helpers.sh` in the `inline` script works (not that the buildpack root is the app dir). And that example _just works_ because the working dir is the app dir.

Reply:
addressed in f1b6564